arbor/include/arbor/mechanism.hpp

```diff
@@ -3,84 +3,109 @@
     using index_type = fvm_index_type;
     using size_type  = fvm_size_type;
 
+    mechanism(const arb_mechanism_type m, const arb_mechanism_interface& i): mech_{m}, iface_{i} {}
     mechanism() = default;
     mechanism(const mechanism&) = delete;
 
     // Return fingerprint of mechanism dynamics source description for validation/replication.
-    virtual const mechanism_fingerprint& fingerprint() const = 0;
+    const mechanism_fingerprint fingerprint() const { return mech_.fingerprint; };
 
     // Name as given in mechanism source.
-    virtual std::string internal_name() const { return ""; }
+    std::string internal_name() const { return mech_.name; }
 
     // Density or point mechanism?
-    virtual mechanismKind kind() const = 0;
+    arb_mechanism_kind kind() const { return mech_.kind; };
 
     // Does the implementation require padding and alignment of shared data structures?
-    virtual unsigned data_alignment() const { return 1; }
+    unsigned data_alignment() const { return iface_.alignment; }
 
-    // Memory use in bytes.
-    virtual std::size_t memory() const = 0;
-
-    // Width of an instance: number of CVs (density mechanism) or sites (point mechanism)
-    // that the mechanism covers.
-    virtual std::size_t size() const = 0;
-
-    // Cloning makes a new object of the derived concrete mechanism type, but does not
-    // copy any state.
-    virtual mechanism_ptr clone() const = 0;
+    // Make a new object of the derived mechanism type, but does not copy any state.
+    mechanism_ptr clone() const { return std::make_unique<mechanism>(mech_, iface_); }
 
     // Non-global parameters can be set post-instantiation:
-    virtual void set_parameter(const std::string& key, const std::vector<fvm_value_type>& values) = 0;
+    void set_parameter(const std::string&, const std::vector<arb_value_type>&);
+
+    // Forward to interface methods
+    void initialize()     { ppack_.vec_t = *time_ptr_ptr; iface_.init_mechanism(&ppack_); }
+    void update_current() { ppack_.vec_t = *time_ptr_ptr; iface_.compute_currents(&ppack_); }
+    void update_state()   { ppack_.vec_t = *time_ptr_ptr; iface_.advance_state(&ppack_); }
+    void update_ions()    { ppack_.vec_t = *time_ptr_ptr; iface_.write_ions(&ppack_); }
+    void post_event()     { ppack_.vec_t = *time_ptr_ptr; iface_.post_event(&ppack_); }
+    void deliver_events(arb_deliverable_event_stream& stream) {
+        ppack_.vec_t  = *time_ptr_ptr;
+        ppack_.events = stream;
+        iface_.apply_events(&ppack_);
+    }
 
     // Peek into state variable
-    virtual fvm_value_type* field_data(const std::string& var) = 0;
+    fvm_value_type* field_data(const std::string& var);
 
-    // Simulation interfaces:
-    virtual void initialize() {};
-    virtual void update_state() {};
-    virtual void update_current() {};
-    virtual void deliver_events() {};
-    virtual void post_event() {};
-    virtual void update_ions() {};
+    mechanism_field_table field_table();
```

Comment:
We have `mechanism::field_table()` here, and `backend::mechanism_field_table()` too. But much like `set_parameter`, this access function could be moved to `shared_state`, avoiding the need to keep copies of stuff in the `mechanism` class. For the GPU mechanism lookup, the pointers to fields could either be copied back over to the CPU side on demand (afaik, we only ever need this to set up probe pointers in fvm_lowered_cell_impl initialization), or a CPU-side copy could be kept in the shared_state per-mechanism data.